shell/common/shell.cc

```diff
@@ -437,6 +451,8 @@ Shell::~Shell() {
       task_runners_.GetIOTaskRunner());
 
   vm_->GetServiceProtocol()->RemoveHandler(this);
+  resource_cache_limit_calculator_->RemoveResourceCacheBytes(
+      reinterpret_cast<uintptr_t>(this));
```

Comment:
I think this approach is fine except for this casting.  A `void*` would be better.  Or if we wanted we could make an interface with zero methods that `Shell` could implement.

Reply:
This is where I was thinking we would just want a weak pointer to the shell. We can check if we are able to lock the weak pointer when we go to iterate the collection, if so we use some virtual method on it to get the resource cache bytes it would want, if not we discard it.

Reply:
@gaaclarke 
You mean change it to this https://github.com/flutter/engine/pull/32156/commits/8e5a23587724a7f4191a74cf960bf7d0e63abc43, right? 

@dnfield 
I guess that when the lightweight engine was designed, the shells were deliberately isolated from each other. All features that require shared state are implemented through shared components, such as `parent_merger` and `parent_io_manager`.  Probably @gaaclarke  can tell us the back story.

Reply:
The reason I don't like a weak_ptr to a Shell is that it would provide the Shell interface to the ResourceCacheLimitCalculator, it doesn't need that kind of power.  Plus it's a heavy interface.

The `void*` looks good to me.  People aren't supposed to touch it or call any methods on it, it's just a handle.

Reply:
I didn't mean to suggest that we'd give it a typed pointer to the shell, but some kind of delegating interface.

void* should generally be avoided in C++ code...

Reply:
@dnfield if you don't like `void*` we should do the interface with no methods.  That's cleaner.  In my mind it was a bit more complication for almost the same usability.

```c++
class ResourceCacheLimitCalculatorItem {
  virtual ~ResourceCacheLimitCalculatorItem() = default;
};

class Shell : public ResourceCacheLimitCalculatorItem {
};

class ResourceCacheLimitCalculator {
  std::map<ResourceCacheLimitCalculatorItem*, int> foo_;
};

```

Reply:
(also avoiding `void*` isn't actually part of the google c++ style)

Reply:
Why wouldn't that interface just have a method on it that says "GetMaxResourceCacheLimit"?

Reply:
That would be fine with me.  It doesn't seem necessary to me.

Reply:
> Why wouldn't that interface just have a method on it that says "GetMaxResourceCacheLimit"?

I also feel that it's not necessary, it complicates things. It would require the shell itself to store catche bytes in a member variable. And I'm not clear on the benefits of doing so.

Reply:
@dnfield  if you don't like `void*`, how about interface `ResourceCacheLimitCalculatorItem` with no method mentioned by @gaaclarke 

Or you just don't like to use `void*` directly? maybe we can use `using`
```
using ResourceCacheLimitCalculatorItem = void*;

class ResourceCacheLimitCalculator {
  std::map<ResourceCacheLimitCalculatorItem, int> foo_;
};
```